sycl/doc/SYCL2020-SpecializationConstants.md

```diff
@@ -0,0 +1,841 @@
+  kernel_single_task(const KernelType &KernelFunc, kernel_handler kh) {
+    KernelFunc(kh);
+  }
+```
+
+It should be transformed into something like this:
+```
+__kernel void oclKernel(args_for_lambda_init, ..., specialization_constants_buffer) {
+  KernelType LocalLambdaClone = { args_for_lambda_init }; // We already do this
+  kernel_handler LocalKernelHandler;
+  LocalKernelHandler.__init_specialization_constants_buffer(specialization_constants_buffer);
+  // for simplicity we could have just used
+  // kernel_handler LocalKernelHandler = { args_for_kernel_handler_init };
+  // here, but we assume that kernel_handler might be used for more than just
+  // accessing specialization constants and therefore there could be other
+  // initialization parameters which also could be conditional
+  // Even now we don't need to always initialize the kernel_handler object
+  // Re-used body of "sycl_kernel" function:
+  {
+     LocalLambdaClone(LocalKernelHandler);
+   }
+}
+```
+
+As mentioned above, creation of `specialization_constants_buffer` kernel
+argument and initialization of `LocalKernelHandler` object with it only happens
+if we are not targeting SPIR-V, i.e. when we compile code for a target without
+native support for specialization constants.
+
+If that new argument was added, it is communicated to DPC++ through regular
+integration header mechanism, i.e. it is added as new entry to
+`kernel_signatures` structure there with parameter kind set to a new
+enumeration value `kernel_param_kind_t::kind_specialization_constants_buffer`.
+
+Those were descriptions of tasks (1) and (2) of DPC++ FE. Task (3) is to help
+DPC++ RT to connect user-provided `specialization_id` variable with
+corresponding symbolic ID of a specialization constant when
+`handler::set_specialization_constant` is invoked.
+
+As noted above, we can't use regular integration header here, because in general
+case, `specialization_id` variables can't be forward-declared. Therefore, we are
+using integration footer approach, which for the following code snippet:
+```
+struct A {
+  float a, b;
+};
+
+specialization_id<int> id_int;
+specialization_id<A> id_A;
+// ...
+[&](handler &cgh) {
+  cgh.set_specialization_constant<id_int>(42);
+  cgh.set_specialization_constant<id_A>({3.14, 3.14});
+}
+```
+
+Will look like:
+
+```
+namespace detail {
+// generic declaration
+template<auto &SpecName>
+struct get_symbolic_id_helper{};
+
+// specializations for each specialization constant:
+// we can refer to all those specialization_id variables, because integration
+// footer was _appended_ to the user-provided translation unit
+template<>
+struct get_symbolic_id_helper<id_int> {
+  static const char *get_symbolic_id() {
+    return "result of __builtin_unique_ID(id_int) encoded here";
+  }
+};
+
+template<>
+struct get_symbolic_id_helper<id_A> {
+  static const char *get_symbolic_id() {
+    return "result of __builtin_unique_ID(A) encoded here";
+  }
+};
+
+} // namespace detail
```

Comment:
The description at the top of this document shows a different content for the integration footer:

```
namespace detail {
  // assuming user defined and used the following specialization_id:
  // constexpr specialiation_id<int> int_const;
  // class Wrapper {
  // public:
  //   static constexpr specialization_id<float> float_const;
  // };

  template<>
  const char *get_spec_constant_symbolic_ID<int_const>() {
    return "unique_name_for_int_const";
  }
  template<>
  const char *get_spec_constant_symbolic_ID<Wrapper::float_const>() {
    return "unique_name_for_Wrapper_float_const";
  }
}
```

I think the two descriptions should agree.  FWIW, I don't see any value for creating the `get_symbolic_id_helper` wrapper struct.  It seems like a simple templated function like `get_spec_constant_symbolic_ID()` would work just fine.

Reply:
Fair enough, removed the struct and aligned those sections in edb1586bd90efed3a8d2bff66abb1e3184bb0168